File: solutions/42.trapping-rain-water.java
/*
 * @lc app=leetcode id=42 lang=java
 *
 * [42] Trapping Rain Water
 */

// @lc code=start


class Solution {
    public int trap(int[] height) {
        int res = 0;
        int leftPos = 0;
        int rightPos = height.length - 1;
        int leftMaxHeight = height[leftPos];
        int rightMaxHeight = height[rightPos];

        while (leftPos < rightPos) {
            int borderHeight = Math.min(leftMaxHeight, rightMaxHeight);
            int currentHeight = 0;
            if (leftMaxHeight <= rightMaxHeight) {
                currentHeight = height[++leftPos];
                leftMaxHeight = Math.max(leftMaxHeight, height[leftPos]);
            } else {
                currentHeight = height[--rightPos];
                rightMaxHeight = Math.max(rightMaxHeight,  height[rightPos]);
            }
            int water = borderHeight - currentHeight;
            res += water > 0 ? water : 0;
        }

        return res;
    }
}
// @lc code=end

File: solutions/42.trapping-rain-water.py
#
# @lc app=leetcode id=42 lang=python3
#
# [42] Trapping Rain Water
#

# @lc code=start
from typing import List


class Solution:
    def trap(self, height: List[int]) -> int:
        res = 0
        left_pos = 0
        right_pos = len(height) - 1
        left_max_height = height[left_pos]
        right_max_height = height[right_pos]

        while left_pos < right_pos:
            border_height = min(left_max_height, right_max_height)
            current_height = 0
            if left_max_height <= right_max_height:
                left_pos += 1
                current_height = height[left_pos]
                left_max_height = max(left_max_height, height[left_pos])
            else:
                right_pos -= 1
                current_height = height[right_pos]
                right_max_height = max(right_max_height, height[right_pos])
            water = border_height - current_height
            res += water if water > 0 else 0

        return res

# @lc code=end

